File: Data Structures & Algorithms/daily-temperatures/submission-2.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> result(temperatures.size(), 0);
        stack<pair<int, int>> stack;

        for (int idx = 0; idx < temperatures.size(); idx++) {
            int curr_temp = temperatures[idx];

            while (!stack.empty() && curr_temp > stack.top().first) {
                auto pair = stack.top();
                stack.pop();
                result[pair.second] = idx - pair.second;
            }


            stack.push({curr_temp, idx});
        }
        return result;
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
            
        first = self.rob_helper(nums[0: -1], 0, {})
        second = self.rob_helper(nums[1:], 0, {})

        return max(first, second)

    def rob_helper(self, nums, idx, cache):
        if idx >= len(nums):
            return 0

        if idx in cache:
            return cache[idx]

        cache[idx] = max(nums[idx] + self.rob_helper(nums, idx + 2, cache), self.rob_helper(nums, idx + 1, cache))
        return cache[idx]
        